ENT-6421 Expose `dumpCheckpoints` on `CordaRPCOps`

lankydan opened the pull request:
Put `dumpCheckpoints` and `debugCheckpoints` on the public API by adding
it to `CordaRPCOps`.

Remove `FlowManagerRPCOps` as its internal functions now live in
`CordaRPCOps`.

vkolomeyko commented:
I wonder if we need a code change at all in this case...
In the Jira I read: "The client wants to dump checkpoints in production for diagnosis purpose for SDK support team to be troubleshoot stuck flows."

Given that required operations are available using `FlowManagerRPCOps` interface, why the client cannot create a Java application which will be using: `MultiRPCClient<FlowManagerRPCOps>` ?

There is already a test `net.corda.node.multiRpc.MultiRpcClientTest#can connect to custom RPC interface` which shows correct API usage.

Also, there is customer-facing documentation on `MultiRPCClient`: https://docs.r3.com/en/platform/corda/4.8/enterprise/node/operating/clientrpc.html#building-the-multi-rpc-client

FAO: @mnesbit

lankydan commented:
Oooh, I didn't realise `MultiRPCClient` was in OS as well.

That is a good point, because if we're making this change because of requirements for ENT users, then they can already do it using `FlowRPCOps` which exists in ENT. If that is good enough, then Viktor is right, we don't need to change anything.

If that is not ok, then I think making `FlowManagerRPCOps` api (moving it out of an internal package) would be enough.

@rick-r3 @adelel1

vkolomeyko commented:
I would even say that it is normal for `FlowManagerRPCOps` to remain `internal`. Those guys who are using it, need to clearly realise that they are crossing the boundary of public API.
On the flip side, if we make `FlowManagerRPCOps` public, the next thing people might ask is backward compatibility of checkpoints JSon output that we produce. I am not sure this is the avenue we would want to explore.

lankydan commented:
That's a fair point.

There is a chance we'll be making more changes around this sort of area in the near future from what I understand, but either way we need some input from @rick-r3, @mnesbit and @adelel1 on what to do here.

mnesbit commented:
Yes I worry that this change will break promises about mixed version support between rpc client and server, but we need to check

lankydan commented:
Discussed with @mnesbit, since we have access to `MultiRPCClient` and we want to make the change as backwards compatible as possible, a new `RPCOps` interface will be added (something like `CheckpointRPCOps`, maybe even `CheckpointDumperRPCOps`) to allow us to make a public version of the current contents of `FlowManagerRPCOps`. `FlowManagerRPCOps` will then be left untouched to ensure backwards compatibility with any clients or shells currently using it.

lankydan committed:
ENT-6421 Public version of `FlowManagerRPCOps`

A public version of `FlowManagerRPCOps` which does not live in an
internal package has been added. This new interface shares the same name
as the internal one.

Because of the name sharing, the internal version has been
`@Deprecated`.

`FlowManagerRPCOpsImpl` implements both the new and old interfaces. This
 allows for backwards compatibility, allowing old shells or clients to
 call the old interface on newer nodes without breaking.

lankydan commented:
I chose to stick with the `FlowManagerRPCOps` naming and therefore deprecated the internal version of the same name. This means we can use this interface to add other flow controlling operations to open source if we choose to do so.

mnesbit commented:
Yes that looks much more sensible

vkolomeyko commented on core/src/main/kotlin/net/corda/core/internal/messaging/FlowManagerRPCOps.kt:
```diff
@@ -5,13 +5,20 @@ import net.corda.core.messaging.RPCOps
 /**
  * RPC operations to perform operations related to flows including management of associated persistent states like checkpoints.
  */
+@Deprecated(
```
I wonder if it would make sense to delete this interface, given that we now have a public version of it.

mnesbit replied:
used by the shell, so perhaps lets give it a version or two
